### PR TITLE
Add NVTX support for performance analysis

### LINKED ISSUE #137: [FEA] Support profiling with NVTX
**Is your feature request related to a problem? Please describe.**
It would be nice to have profiling support so that we can analyze the performance easily (visualizing the bottleneck through [NVIDIA Nsight Systems](https://developer.nvidia.com/nsight-systems)).

**Describe the solution you'd like**

- Make use of [NVTX](https://github.com/NVIDIA/NVTX).
- Provide a build flag to enable/disable the use of NVTX.
  - Check the overhead of NVTX and enable it by default if the overhead is not much
- Make runtime-configurable through an environment variable or API call.

### PR DESCRIPTION
**Pull request by gigony**
Need to rebase once https://github.com/rapidsai/cucim/pull/141 is merged.

- Instrumented important methods in cuCIM with NVTX so can see performance bottlenecks easily with NSight systems.
- Tracing can be enabled through config file or environment variable or through API and less than 1% performance overheads in normal execution.

#### Through `.cucim.json` file

```json
{
        "profiler" : { "trace": true }
}
```

#### Through Environment variable

```bash
CUCIM_TRACE=1 python
```

#### Through API

```python
from cucim import CuImage

CuImage.profiler(trace=True)
#or
CuImage.profiler().trace(True)

CuImage.profiler().config
# {'trace': True}
CuImage.profiler().trace()
# True
CuImage.is_trace_enabled # this is simpler method.
# True
```

### Profiling with NVIDIA Nsight Systems

```bash
nsys profile -f true -t cuda,nvtx,osrt -s cpu -x true --trace-fork-before-exec true -o my_profile `which python` benchmark.py
# can add `--stats true`
```

Then, execute `nsight-sys` to open the profile results (my_profile.qdrep).

image

closes #137

### CODE REVIEW

**Comment by jakirkham** [comment]
PR ( https://github.com/rapidsai/cucim/pull/141 ) is in. So this is ready for rebasing

**Comment by gigony** [comment]
> PR ( #141 ) is in. So this is ready for rebasing

@jakirkham 
Please help merge this when you have time.

File: cpp/cmake/deps/nvtx3.cmake [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+if (NOT TARGET deps::nvtx3)
+    FetchContent_Declare(
+            deps-nvtx3
+            GIT_REPOSITORY https://github.com/NVIDIA/NVTX.git
+            GIT_TAG 3c98c8425b0376fd8653aac7cfc6a864f3897752
+            # GIT_SHALLOW TRUE  # Uncomment this when the official release of nvtx3-cpp is available.
```
Comment (jakirkham) [inline cpp/cmake/deps/nvtx3.cmake:21]: Should we raise an issue to track this?

File: cpp/cmake/deps/nvtx3.cmake [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    )
+    FetchContent_GetProperties(deps-nvtx3)
+    if (NOT deps-nvtx3_POPULATED)
+        message(STATUS "Fetching nvtx3 sources")
+        FetchContent_Populate(deps-nvtx3)
+        message(STATUS "Fetching nvtx3 sources - done")
+    endif ()
+
+    # Create shared library
+    cucim_set_build_shared_libs(ON) # since nvtx3 is head-only library, this may not needed.
```
Comment (jakirkham) [inline cpp/cmake/deps/nvtx3.cmake:31]:
```suggestion
    cucim_set_build_shared_libs(ON) # since nvtx3 is header-only library, this may not needed.
```

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Had a couple questions above. Otherwise LGTM

**Review by jakirkham — APPROVED** [review]

File: cpp/cmake/deps/nvtx3.cmake (hunk 1 above)
Comment (gigony) [reply cpp/cmake/deps/nvtx3.cmake]: Thanks @jakirkham!

Created #168 and updated the comment in `nvtx3.cmake`.

**Review by jakirkham — APPROVED** [review]

**Review by ajschmidt8 — APPROVED** [review]

**Comment by ajschmidt8** [comment]
CI is passing. Merging.

**Comment by jakirkham** [comment]
Thanks AJ! 😄

**Pull request merged**